### PR TITLE
Cleanup

### PR DESCRIPTION
**Pull request by canndrew**
This refactors the library somewhat.

Major changes are:
- Split the `IfAddr` type into a few types.
- Add a dependency on the `c_linked_list` crate.

### CODE REVIEW

**Comment by maidsafe-highfive** [comment]
r? @maqi

(maidsafe_highfive has picked a reviewer for you, use r? to override)

File: src/lib.rs [hunk 1]
```diff
@@ -33,47 +33,95 @@
         unconditional_recursion, unknown_lints, unsafe_code, unused, unused_allocation,
         unused_attributes, unused_comparisons, unused_features, unused_parens, while_true)]
 #![warn(trivial_casts, trivial_numeric_casts, unused_extern_crates, unused_import_braces,
-        unused_qualifications, unused_results, variant_size_differences)]
+        unused_qualifications, unused_results)]
```
Comment (maqi) [inline src/lib.rs:36]: any particular reason this warning being muted?

File: src/lib.rs [hunk 2]
```diff
@@ -304,15 +371,15 @@ mod getifaddrs_windows {
...
+    pub fn get_if_addrs() -> io::Result<Vec<Interface>> {
+        let mut ret = Vec::<Interface>::new();
         let mut ifaddrs: *const IpAdapterAddresses;
         let mut buffersize: c_ulong = 15000;
         loop {
             unsafe {
                 ifaddrs = libc::malloc(buffersize as size_t) as *mut IpAdapterAddresses;
                 if ifaddrs.is_null() {
-                    panic!("Failed to allocate buffer in getifaddrs()");
+                    panic!("Failed to allocate buffer in get_if_addrs()");
```
Comment (maqi) [inline src/lib.rs:382]: shall here be consistent as other places to return an error?

File: src/lib.rs [hunk 3]
```diff
@@ -332,179 +399,136 @@ mod getifaddrs_windows {
...
-
-fn is_loopback(ip_addr: &IpAddr) -> bool {
-    match *ip_addr {
-        IpAddr::V4(a) => a.octets()[0] == 127,
-        IpAddr::V6(a) => a.segments() == [0, 0, 0, 0, 0, 0, 0, 1],
-    }
-}
-
-/// Remove loopback address(s)
-pub fn filter_loopback(mut ifaddrs: Vec<IfAddr>) -> Vec<IfAddr> {
```
Comment (maqi) [inline src/lib.rs:494]: This function has now been removed, does it mean it's now all the up-level code's responsibility to filter loopback by calling is_loopback function?

File: src/lib.rs (hunk 1 above)
Comment (canndrew) [inline src/lib.rs:36]: Because it's triggered by the `IfAddr` type. It doesn't seem like much of an issue though.

File: src/lib.rs (hunk 2 above)
Comment (canndrew) [inline src/lib.rs:382]: I'm on the fence about this one. As much as I dislike `panic!` it's standard behaviour for Rust code to panic on out-of-memory.

File: src/lib.rs (hunk 3 above)
Comment (canndrew) [inline src/lib.rs:494]: That's right. I added an `is_loopback` method to the `Interface`, `IfAddr`, `Ifv4Addr` and `Ifv6Addr` types. If someone's iterating over interfaces and they want to filter out the loopback interfaces they can call `.filter(|i| !i.is_loopback())`. This seems a bit clearer and a bit more normal to me.

**Pull request merged**